### PR TITLE
prov/efa: add a document that describes EFA RDM protocol v4

### PR DESCRIPTION
**Pull request by wzamazon**
This patch added a document that describes EFA RDM protocol version
4. The document is written in markdown format.

Signed-off-by: Wei Zhang <wzam@amazon.com>

### CODE REVIEW

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
Changes:

1. fixed the doc according to comments from @corey-a-henderson, many thanks!
2. added a commit to rename some protocols.
3. added a section to describe emulated atomic sub-protocols (I forget to add in the first version)
4. added the drawio file I used to generate the png file.

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 1]
```diff
@@ -0,0 +1,1318 @@
...
+Table: 1.2 a list of packet type IDs
+
+| Packet Type ID  | Nick Name   | Full Name                 | Used by                       |
+|-|-|-|-|
+| 1               | RTS         | Request To Send           | Deprecated                    |
+| 2               | CONNACK     | CONNection ACKnowlegement | Deprecated                    |
+| 3               | CTS         | Clear To Send             | long-CTS message/read/write |
+| 4               | DATA        | Data                      | long-CTS message/read/write |
+| 5               | READRSP     | READ ReSPonse             | emulated short/long-read      |
+| 6               | N/A         | N/A                       | reserved for internal use      |
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:155]: Is this reserved or just not used?

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 2]
```diff
@@ -0,0 +1,1318 @@
...
+| 7               | EOR         | End Of Read               | long-read message/write     |
+| 8               | ATOMRSP     | ATOMic ResSPonse          | emulated write/fetch/compare atomic |
+| 9               | HANDKSHAKE  | Handshake                 | handshake                     |
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:158]: typo in nickname

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 3]
```diff
@@ -0,0 +1,1318 @@
...
+- Emulated fetch atomic
+- Emulated compare atomic
+
+The definition of these baseline features are in chapter 3. Any endpoint that adopts protocol
+v4 must support these baseline features.
+
+Protocol v4 then allow changes to be introduced as "extra feature" and "extra request".
+Each extra feature/request will be assigned an ID, when it is introduce to protocol v4.
+The ID starts from 0, and increases by 1 for each extra feature/request. Typically,
+new extra feature/request is introduced with libfaric minor releases, and will NOT be
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:237]: typo here  - libfabric

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 4]
```diff
@@ -0,0 +1,1318 @@
...
+| `nex_p3`  | 4 | integer | `uint32_t` |
+| `exinfo`  | `8 * (nex_p3 - 3)` | integer array | `uint64_t[]` |
+
+The first 4 bytes (3 fields: `type`, `version`, `flags`) is the EFA RDM base header (section 1.3).
+
+Immediately after the base header, there are 2 fields `nex_p3` and `exinfo`. 
+
+The field `exinfo` is an array of 8 byte integers, which stores the capability of an endpoint.
+
+As mentioned before, each extra feature/request was assigned an ID when it was introduced to protocol v4.
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:318]: This is just a bitmap, right? I think I'd call it that in the type field above and simplify this paragraph.

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 5]
```diff
@@ -0,0 +1,1318 @@
...
+toggle on the No. 0 (the first) bit of `exinfo[0]`. (section 4.1)
+
+If an endpoint wants to impose the "stable header length" extra request it need to toggle on bit No 2.
+in `exinfo[0]`. (section 4.3)
+
+Note, the field `exinfo` was named `features` when protocol v4 was initially introduced, at that time we
+only planned for extra features. Later, we discovered that the handshake sub-protocol can also be used to pass
+additional request information, thus introduced the concept of "extra request" and renamed this field `exinfo`.
+
+`nex_p3` is number of `exinfo` flags of the endpoint plus 3. The "plus 3" is for historical reasons.
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:334]: Historical reasons? Or backwards compatibility? I'm a bit lost here with the changes to this packet.

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 6]
```diff
@@ -0,0 +1,1318 @@
...
+
+### 3.2 baseline features for two-sided communication
+
+This section describes the 3 baseline features for two sided communication: eager message, medium message, long-CTS message.
+Each of them correspond to the same named sub-protocol. When describing a sub-protocol, we always follow
+the same structure: workflow, packet format and implementation tips.
+
+#### Eager message feature/sub-protocol
+
+Eager message feature/sub-protocol is used when application's send buffer is small enough to be fit in one packet.
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:517]: "small enough to fit"

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 7]
```diff
@@ -0,0 +1,1318 @@
...
+This protocol works in the following order:
+
+1. On sender side, application call libfabric's send API, providing a send buffer.
+2. On receiver side, application call libfabric's receive API, providing a receive buffer.
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:521]: Ideally 1 and 2 are flipped here since this order could lead to an unexpected message

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 8]
```diff
@@ -0,0 +1,1318 @@
...
+3. Sender sends an EAGER_RTM (EAGER_MSGRTM or EAGER_TAGRTM) packet, which contains the application's data.
+4. Upon receiving the packet, receiver will process the received packet, and make sure the received
+   data is in application's receive buffer.
+
+The following diagram illustrate the workflow:
+
+![eager message](message_eager.png)
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:528]: I'd simplify the language in these diagrams to just "Send completion" and "Receive completion"

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 9]
```diff
@@ -0,0 +1,1318 @@
...
+is the eager RTM packet arrived before application called libfabric's receive API.
+
+4. if application does not require ordered send, it would be possible to use application's receive buffer to receive data
+directly. In this case, receiver might need the sender to keep the packet header length constant through out the communication.
+The extra request "stable header length" is designed for this use case, see chapter 4.3 for more discussion on this topic.
+
+5. One might notice that there is no application data length in the header, so how can the receiver of an eager RTM packet
+   know how many application data is in the packet? The answer is to use the following formula:
+
+        application_data_length = total_packet_size - RTM mandatoary header length - REQ optional header length
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:567]: typo - mandatory

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 10]
```diff
@@ -0,0 +1,1318 @@
...
+There are 3 fields that is new:
+
+`msg_length` is the length of the whole application message.
+
+`send_id` is an ID the sending endpoint assigned to the send operation, and receive should include
+`send_id` in CTS packet. An endpoint will have multiple send operations at the same time, thus
+when processing a CTS packet from a receive, it needs a way to locate the send operation the
+CTS packet is referring to.
+
+Admittedly, the introduction of `send_id` is not absolute necessary, because receiver could have
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:687]: absolutely

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 11]
```diff
@@ -0,0 +1,1318 @@
...
+included `msg_id` in CTS header, and sender should be able to locate the send operation using
+the combination of receiver's address and message ID. However, that approach would require
+the sending endpoint set up a map between (address + `msg_id`) and send operation, and look up the map every time
+it received a CTS packet. We considered that approach too burdensome for an endpoint to implmement
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:691]: implement

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 12]
```diff
@@ -0,0 +1,1318 @@
...
+
+The 3 new fields in the header are `pad`, `recv_id` and `recv_length`.
+
+The field `pad` is 4 byte padding to make the variables align on 8 bytes boundary.
+
+The field `recv_id` is similar to `send_id` introduced earlier, but for an on-going receive operation.
+Sender should include `recv_id` in the DATA packet.
+
+The field `recv_length` is the number of bytes receiver is ready to receive for this operation,
+it must be > 0 to make the communication going forward.
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:731]: "it must be >0 otherwise the receiver should wait to post the CTS packet" might be clearer here

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 13]
```diff
@@ -0,0 +1,1318 @@
...
+
+Before getting into details of each feature, we want to discuss some topics related to one-sided operation.
+
+There are 3 types of one-sided operations: write, read and atomic.
+
+Like in two-sided communcation, there are also two endpoints involved in one-sided communcation.
+However, only on one side will application call libfabric's one-sided API (such as `fi_write`,
+`fi_read` and `fi_atomic`). In protocol v4, this side is called requester.
+
+On the other side (which is called responder), application does not make calls to lifabric API call,
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:772]: libfabric

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 14]
```diff
@@ -0,0 +1,1318 @@
...
+| `type`           | 1 | integer | `uint8_t`  | part of base header |
+| `version`        | 1 | integer | `uint8_t`  | part of base header|
+| `flags`          | 2 | integer | `uint16_t` | part of base header |
+| `msg_id`         | 4 | integer | `uint32_t` | message ID |
+| `msg_length`     | 4 | integer | `uint64_t` | total length of the message |
+| `send_id`        | 4 | integer | `uint32_t` | ID of the receive operation  |
+| `read_iov_count` | 4 | integer | `uint32_t` | number of iov to read |
+| `read_iov`       | `read_iov_count` * 24 | array | `efa_rma_iov[]` | read iov information |
+
+In LONGREAD_RTM packet mandatory header, there are two new fields `read_iov_count` and `read_iov`.
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:1057]: This is probably too specific for a protocol doc, instead maybe just state that read_iov needs contain the right descriptor so that RDMA read can be used to retrieve the data from the sender's buffer?

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 15]
```diff
@@ -0,0 +1,1318 @@
...
+| Name | Length (bytes) | type | C language type | Note |
+|-|-|-|-|-|
+| `type`           | 1 | integer | `uint8_t`  | part of base header |
+| `version`        | 1 | integer | `uint8_t`  | part of base header|
+| `flags`          | 2 | integer | `uint16_t` | part of base header |
+| `send_id`        | 4 | integer | `uint32_t` | ID of the send operation |
+| `recv_id`        | 4 | integer | `uint32_t` | ID of the receive operation |
+| `connid`         | 4 | integer | `uint32_t` | optional sender connection ID, see section 4.4 for more detail|
+
+#### emulated long-read write sub-protocol
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:1089]: Is there a better name for this? It's kind of hard to parse although I know what you're talking about here. Maybe "emulated long RDMA write via RDMA read"?

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 16]
```diff
@@ -0,0 +1,1318 @@
...
+Direct read sub-protocol is the simplest sub-protocol in protocol v4. It does not involve a REQ packet.
+The workflow is just for read requester keep using RDMA read on the responder. For this protocol, it is
+not necessary that responder keep progress engine running.
+
+### 4.2 delivery complete
+
+The extra feature "delivery complete" was introduced with libfabric 1.12.0, and was assigned ID 1.
+
+Delivery complete is a requirement application can impose on an endpoint when opening the endpoint.
+It requires that when application got the send/write completion, the application data must have
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:1135]: "when an application gets a send/write completion"

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 17]
```diff
@@ -0,0 +1,1318 @@
...
+Because of that, it is possible for an endpoint to receive packets from a destroyed
+peer. The endpoint does not know how to handle such a packet, and will crash or cause data corruption.
+
+The problem can be demonstrated using the following workflow:
+
+    1. We have two instances A and B.
+    2. On instance A we create endpoint 1.
+    3. On instance B we create endpoint 2.
+    4. Endpoint 1 and 2 start communication with each other.
+    5. There is an accident on instance B, endpoint 2 was destroyed.
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:1271]: "Instance B crashes, endpoint 2 is destroyed"

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 18]
```diff
@@ -0,0 +1,1318 @@
...
+    6. Before it is destroyed, endpoint 2 sent some packets to 1.
+    7. On instance B, a new endpoint 3 was created, which has the same QPN as endpoint 2.
+    8. Endpoint 3 start sending packets to endpoint 1.
+    9. Upon receive packets from endpoint 3, endpoint 1 recognize this is a new peer via its connection ID.
+       and will insert endpoint 3's address to address vector.
+    10. The packets from endpoint 2 (in step 6) arrived, because endpoint 2 has the same gid+qpn as endpoint 3.
+        EFA device will think these packets are from endpoint 3.
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:1278]: This isn't really a device problem, so I'd say EFA provider instead.

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 19]
```diff
@@ -0,0 +1,1318 @@
...
+    11. Endpoint 1 will not be able to process these packets, and will crash.
+
+The solution to this problem is to include sender's connid in each packet's header. In this case,
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:1281]: Does it make sense to put connid in the base header instead so we don't have to add it to each packet type (now and in the future)? Or is it simpler to not do that?

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 20]
```diff
@@ -0,0 +1,1318 @@
...
+aspects of the protocol, but did not give specific instruction about what implementation should do.
+
+There are a few things that this document consider as implementation specific thus does not cover.
+
+For example, this document does not specify the selection logic of various protocols. For example, there are
+three message sub-protocols medium message, long-CTS message and long-read message that can be used for
+long message (it is not recommended), and an implementation can choose switch point without breaking
+protocol.
+
+On similar note, this document does not describe the shared memroy (SHM) implementation of EFA provider, which is an optimization technique
```
Comment (rwespetal) [inline prov/efa/docs/efa_rdm_protocol_v4.md:1310]: memory

**Review by rwespetal — COMMENTED** [review]
Thanks, Wei. I took a quick pass at the protocol doc and will scan the code changes later.

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 21]
```diff
@@ -0,0 +1,1318 @@
...
+this document because they are considered "implementation details".
+
+## 1. Basics
+
+EFA RDM communication protocol is for two lifabric endpoints to use EFA device to communication
+with each other.
+
+### 1.1 Why is EFA RDM communication protocol needed?
+
+The reason we need a EFA RDM communication protocol is because of EFA device's limited
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md:75]: Fixed in new revision.

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 22]
```diff
@@ -0,0 +1,1318 @@
...
+capability. Currently, EFA device supports the following two types of communications:
+
+ 1. send/receive a message up to EFA device's Maximum Transmission Unit (MTU) size.
+ 2. RDMA read of a memory buffer up to 1GB (if both endpoints' software stacks support RDMA read).
+ 
+Moreover, for send/receive, EFA device does not guarantee ordered delivery, e.g. when sender
+sends multiple messages to a receiver, the receiver may receive the packets in an order different
+from how they are sent.
+
+Protocol v4 defines how two endpoints can use EFA device's limited capability to achieve:
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md:85]: remove the word `limited` in new revision.

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 23]
```diff
@@ -0,0 +1,1318 @@
...
+To meet application's specific needs, protocol v4 defines a set of sub-protocols for
+as listed in table 1.1:
+
+Table: 1.1 a list of sub-protocols
+
+| Sub Protocol Name             | Used For  | Definition in |
+|-|-|-|
+| Eager message                 | Two sided | Section 3.2   |
+| Medium message                | Two sided | Section 3.2   |
+| Long-CTS message              | Two sided | Section 3.2   |
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md:106]: long-read is also a rendezvous protocol. The one unique thing about this protocol is the use of CTS packet, thus I renamed it long-CTS.

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 24]
```diff
@@ -0,0 +1,1318 @@
...
+information between two endpoints.
+
+A packet is a message that does not exceed MTU size, and is exchanged between
+two endpoints using EFA device's send/receive capability.
+
+Protocol v4 defines a set of packet types. They can be split into two category:
+REQ packet types and non-REQ packet types.
+
+Typically, in each sub-protocol, one endpoint send a REQ packet to the other to initialize
+the communication. Each sub-protocol is unique, thus each sub-protocol defines its own REQ
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md:136]: That was the purpose of the new 2 paragraphs. I will rephrase them.

File: prov/efa/docs/efa_rdm_protocol_v4.md (hunk 16 above)
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md:1135]: fixed

File: prov/efa/docs/efa_rdm_protocol_v4.md (hunk 17 above)
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md:1271]: I added `application crashed, hence endpoint 2 was destroyed`.  The instance itself might be OK in this case.

File: prov/efa/docs/efa_rdm_protocol_v4.md (hunk 18 above)
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md:1278]: fixed

File: prov/efa/docs/efa_rdm_protocol_v4.md (hunk 19 above)
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md:1281]: I think it is simpler to add it to each packet type, and using a bit in `flags` to indicate its existence in the header.

To add it to base header would require introduce new packet types. Because current packet header are all based on 4 bytes base header, and the alignment would change.

Comment (rwespetal) [reply prov/efa/docs/efa_rdm_protocol_v4.md:1283]: Got it, makes sense.

File: prov/efa/docs/efa_rdm_protocol_v4.md (hunk 20 above)
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md]: fixed in new revision

**Comment by wzamazon** [comment]
Closed in favor of https://github.com/ofiwg/libfabric/pull/6945
